### PR TITLE
Move Derefer before Retag

### PR DESCRIPTION
**Pull request by ouz-a**
_Follow up work to #96116 #95857 #95649_

This moves `Derefer` before `Retag` and creates a new `LocalInfo` called `Temp` to avoid retagging created temp values. 

Zulip discussion [link](https://rust-lang.zulipchat.com/#narrow/stream/189540-t-compiler.2Fwg-mir-opt/topic/deref.20as.20first.20and.20only.20projection)

r? @oli-obk

### CODE REVIEW

File: compiler/rustc_middle/src/mir/mod.rs [hunk 1]
```diff
@@ -1079,6 +1079,8 @@ pub enum LocalInfo<'tcx> {
     /// A temporary created during the creation of an aggregate
     /// (e.g. a temporary for `foo` in `MyStruct { my_field: foo }`)
     AggregateTemp,
+    /// A temporary created during the pass `Derefer` to avoid it's retagging
+    Temp,
```
Comment (ouz-a) [inline compiler/rustc_middle/src/mir/mod.rs:1083]: Not sure about this name

File: compiler/rustc_middle/src/mir/patch.rs [hunk 2]
```diff
@@ -78,6 +78,15 @@ impl<'tcx> MirPatch<'tcx> {
         Location { block: bb, statement_index: offset }
     }
 
+    pub fn new_local_temp(&mut self, ty: Ty<'tcx>, span: Span) -> Local {
```
Comment (ouz-a) [inline compiler/rustc_middle/src/mir/patch.rs:81]: Again needs a better name(?) but I am not sure how to differentiate between new temp temp and new temp

File: compiler/rustc_middle/src/mir/mod.rs (hunk 1 above)
Comment (oli-obk) [reply compiler/rustc_middle/src/mir/mod.rs:1083]: Maybe `DerefTemp`? Just like `AggregateTemp`, but created for references that get immediately dereferenced in place expressions containing dereferences in the middle.

File: compiler/rustc_mir_transform/src/add_retag.rs [hunk 3]
```diff
@@ -57,6 +57,17 @@ fn may_be_reference(ty: Ty<'_>) -> bool {
     }
 }
 
+/// Determines whether or not this LocalDecl is temp, if not it needs retagging.
+fn is_not_temp<'tcx>(local_decl: &LocalDecl<'tcx>) -> bool {
+    if local_decl.local_info.is_some() {
```
Comment (oli-obk) [inline compiler/rustc_mir_transform/src/add_retag.rs:62]: use `if let Some(...` so you don't need to unwrap in the `match` below

**Comment by oli-obk** [comment]
some style nits and CI is unhappy with the `src/test/run-make/const_fn_mir` test

File: compiler/rustc_middle/src/mir/mod.rs [hunk 4]
```diff
@@ -1178,6 +1180,10 @@ impl<'tcx> LocalDecl<'tcx> {
         }
     }
 
+    pub fn temp(mut self) {
```
[the remark this answers is not in the document]
Comment (ouz-a) [reply compiler/rustc_middle/src/mir/mod.rs:1183]: Oh that function doesn't even do anything I thought I deleted it

File: compiler/rustc_middle/src/mir/patch.rs (hunk 2 above)
Comment (ouz-a) [reply compiler/rustc_middle/src/mir/patch.rs]: I completely missed this and the comment above 🤦‍♂️

File: compiler/rustc_middle/src/mir/mod.rs (hunk 1 above)
Comment (ouz-a) [reply compiler/rustc_middle/src/mir/mod.rs]: Sounds better!

File: compiler/rustc_middle/src/mir/mod.rs [hunk 5]
```diff
@@ -1079,6 +1079,8 @@ pub enum LocalInfo<'tcx> {
     /// A temporary created during the creation of an aggregate
     /// (e.g. a temporary for `foo` in `MyStruct { my_field: foo }`)
     AggregateTemp,
+    /// A temporary created during the pass `Derefer` to avoid it's retagging
```
Comment (RalfJung) [inline compiler/rustc_middle/src/mir/mod.rs:1082]: (should be "its retagging"... not a big deal though)

File: compiler/rustc_mir_transform/src/add_retag.rs [hunk 6]
```diff
@@ -57,6 +57,17 @@ fn may_be_reference(ty: Ty<'_>) -> bool {
     }
 }
 
+/// Determines whether or not this LocalDecl is temp, if not it needs retagging.
+fn is_not_temp<'tcx>(local_decl: &LocalDecl<'tcx>) -> bool {
```
[the remark this answers is not in the document]
Comment (ouz-a) [reply compiler/rustc_mir_transform/src/add_retag.rs:61]: You are right, I will change this function in the next update of derefer.

**Pull request merged**